### ISSUE TITLE
Connection aborted.', RemoteDisconnected('Remote end closed connection without response')

### ISSUE THREAD

**Issue by FaresKi** [Azure/azure-kusto-python#478]
#### Code Sample, a copy-pastable example if possible

```python
def execute_and_calculate_query(self, query: str):
        """
        Execute a query and calculate the ratio of the results."""
        field_ratio_values = []
        try:
            query_result = self.kusto_client.execute_query(self.kusto_database, query)
        except KustoServiceError as err:
            logging.exception(
                f"Can't seem to reach database. Timeout error maybe? Will assign 0 for now. Error: {err}"
            )
            return None
        dataframe = dataframe_from_result_table(query_result.primary_results[0])
        return dataframe
```
#### Problem description
For some reason, that I haven't managed to pinpoint yet, the connection is randomly aborted (this only happens when deployed in an environment such as Azure Functions)
It's important to understand that I'm firing a series of queries to my Kusto cluster, with a 20 second gap between each query, and this error comes up randomly:

```
[Error] Traceback (most recent call last):File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 703, in urlopenhttplib_response = self._make_request(File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 449, in _make_requestsix.raise_from(e, None)

File "<string>", line 3, in raise_fromFile "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 444, in _make_requesthttplib_response = conn.getresponse()

File "/usr/local/lib/python3.10/http/client.py", line 1374, in getresponseresponse.begin()

File "/usr/local/lib/python3.10/http/client.py", line 318, in beginversion, status, reason = self._read_status()File "/usr/local/lib/python3.10/http/client.py", line 287, in _read_statusraise RemoteDisconnected("Remote end closed connection without"http.client.RemoteDisconnected: Remote end closed connection without responseDuring handling of the above exception, another exception occurred:Traceback (most recent call last):File "/home/site/wwwroot/.python_packages/lib/site-packages/requests/adapters.py", line 487, in sendresp = conn.urlopen(File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 787, in urlopenretries = retries.increment(File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/util/retry.py", line 550, in incrementraise six.reraise(type(error), error, _stacktrace)File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/packages/six.py", line 769, in reraiseraise value.with_traceback(tb)File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 703, in urlopenhttplib_response = self._make_request(File "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 449, in _make_requestsix.raise_from(e, None)File "<string>", line 3, in raise_fromFile "/home/site/wwwroot/.python_packages/lib/site-packages/urllib3/connectionpool.py", line 444, in _make_requesthttplib_response = conn.getresponse()File "/usr/local/lib/python3.10/http/client.py", line 1374, in getresponseresponse.begin()

File "/usr/local/lib/python3.10/http/client.py", line 318, in beginversion, status, reason = self._read_status()File "/usr/local/lib/python3.10/http/client.py", line 287, in _read_statusraise RemoteDisconnected("Remote end closed connection without"urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))During handling of the above exception, another exception occurred:Traceback (most recent call last):File "/home/site/wwwroot/compute_metrics_service/app.py", line 55, in mainentity_fields = helpers.compute_completeness(File "/home/site/wwwroot/compute_metrics_service/utils/__init__.py", line 130, in compute_completenessresult = self.execute_and_calculate_query(query)File "/home/site/wwwroot/compute_metrics_service/utils/__init__.py", line 37, in execute_and_calculate_queryquery_result = self.kusto_client.execute_query(self.kusto_database, query)File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/tracing/decorator.py", line 76, in wrapper_use_tracerreturn func(*args, **kwargs)

File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/kusto/data/client.py", line 175, in execute_queryreturn self._execute(self._query_endpoint, database, query, None, self._query_default_timeout, properties)

File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/kusto/data/client.py", line 281, in _executeresponse = KustoTracing.call_func_tracing(File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/kusto/data/_telemetry.py", line 118, in call_func_tracingreturn kusto_func(*args, **kwargs)

File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/tracing/decorator.py", line 76, in wrapper_use_tracerreturn func(*args, **kwargs)

File "/home/site/wwwroot/.python_packages/lib/site-packages/requests/sessions.py", line 635, in postreturn self.request("POST", url, data=data, json=json, **kwargs)File "/home/site/wwwroot/.python_packages/lib/site-packages/requests/sessions.py", line 587, in requestresp = self.send(prep, **send_kwargs)

File "/home/site/wwwroot/.python_packages/lib/site-packages/requests/sessions.py", line 701, in sendr = adapter.send(request, **kwargs)

File "/home/site/wwwroot/.python_packages/lib/site-packages/requests/adapters.py", line 502, in sendraise ConnectionError(err, request=request)requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

#### If query related, does it happen on other platforms (Kusto Web UI, Kusto Explorer)? 

Not to my knowledge, no.

#### Output of ``pip freeze`` 

<details>

azure-common==1.1.28
azure-core==1.26.4
azure-data-tables==12.4.2
azure-functions==1.13.3
azure-identity==1.12.0
azure-kusto-data==4.1.4
azure-kusto-ingest==4.1.4
azure-mgmt-core==1.4.0
azure-mgmt-securityinsight==1.0.0
azure-storage-blob==12.16.0
azure-storage-queue==12.6.0
base58==2.1.1
black @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_1f1pcodmuo/croot/black_1680737253550/work
certifi==2022.12.7
cffi @ file:///Users/runner/miniforge3/conda-bld/cffi_1671179414629/work
cfgv @ file:///home/conda/feedstock_root/build_artifacts/cfgv_1629909281805/work
charset-normalizer==3.1.0
click @ file:///opt/concourse/worker/volumes/live/2d66025a-4d79-47c4-43be-6220928b6c82/volume/click_1646056610594/work
coverage==7.2.3
cryptography==40.0.2
distlib @ file:///home/conda/feedstock_root/build_artifacts/distlib_1668356257807/work
exceptiongroup==1.1.1
filelock @ file:///home/conda/feedstock_root/build_artifacts/filelock_1681839547898/work
identify @ file:///home/conda/feedstock_root/build_artifacts/identify_1679693584347/work
idna==3.4
ijson==3.2.0.post0
iniconfig==2.0.0
isodate==0.6.1
msal==1.22.0
msal-extensions==1.0.0
msrest==0.7.1
multidict==6.0.4
mypy-extensions==0.4.3
nodeenv @ file:///home/conda/feedstock_root/build_artifacts/nodeenv_1656187504160/work
numpy==1.24.2
oauthlib==3.2.2
packaging @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_c3zlhpboab/croot/packaging_1678965318579/work
pandas==2.0.0
pathspec @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_b8rj_doxjl/croot/pathspec_1674681570418/work
platformdirs @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_7fs8_2xgrm/croots/recipe/platformdirs_1662711383474/work
pluggy==1.0.0
portalocker==2.7.0
pre-commit @ file:///home/conda/feedstock_root/build_artifacts/pre-commit_1680693009076/work
pycparser @ file:///home/conda/feedstock_root/build_artifacts/pycparser_1636257122734/work
PyJWT==2.6.0
pytest==7.3.1
pytest-cov==4.0.0
python-dateutil==2.8.2
pytz==2023.3
PyYAML @ file:///Users/runner/miniforge3/conda-bld/pyyaml_1666772684107/work
requests==2.28.2
requests-oauthlib==1.3.1
six==1.16.0
tenacity==8.2.2
tomli @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_90762ba4-f339-47e8-bd29-416854a59b233d27hku_/croots/recipe/tomli_1657175507767/work
typing_extensions==4.5.0
tzdata==2023.3
ukkonen @ file:///Users/runner/miniforge3/conda-bld/ukkonen_1666949190900/work
urllib3==1.26.15
virtualenv @ file:///Users/runner/miniforge3/conda-bld/virtualenv_1670705731972/work
yarl==1.8.2

</details>

**Comment by yogilad** [comment]
This sounds like an issue having to do with Azure Functions limiting your runtime.
Do you have any client activity ids for queries that got aborted (+time &  cluster name)?

**Comment by yogilad** [comment]
@FaresKi , did you manage to resolve this?

**Comment by FaresKi** [comment]
> @FaresKi , did you manage to resolve this?

I managed to solve the problem by implementing a retry policy using tenacity's retry library.

**Comment by yogilad** [comment]
Thanks.
Labeling as bug. Need to make sure this exception is caught and rethrown as a KustoClientException with proper context (client requets id, etc.)

**Comment by AsafMah** [comment]
Solved by #486 
Will be released on the next version
